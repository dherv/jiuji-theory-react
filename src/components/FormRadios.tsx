import { Field } from 'formik';
import React, { FC, Fragment } from 'react';
import styled from 'styled-components';
import { themeBlue, themeBlueLight } from '../styled/themes';
import Button from './Button';
import { FormLabel } from './FormCreateEdit';

const FormRadio: FC<{ name: string; data: any[]; selected: string }> = ({
  name,
  data,
  selected,
  ...rest
}) => {
  return (
    <div role="group" aria-labelledby={name}>
      <FormLabel as="div" id={name}>
        {name}
      </FormLabel>
      <FormRadioGroup>
        {data.map((d) => (
          <Fragment key={d.id}>
            <FormRadioButton
              type="button"
              theme={selected === d.name ? themeBlue : themeBlueLight}
            >
              <label
                style={{ display: "block", padding: ".5rem 1rem" }}
                htmlFor={d.name}
              >
                {d.name}
              </label>
            </FormRadioButton>

            <Field
              type="radio"
              value={d.name}
              name={name}
              id={d.name}
              style={{ display: "none" }}
            ></Field>
          </Fragment>
        ))}
      </FormRadioGroup>
    </div>
  );
};

const FormRadioButton = styled(Button)`
  padding: 0;
  margin-right: 1rem;
`;

const FormRadioGroup = styled.div`
  display: flex;
`;

export default FormRadio;
