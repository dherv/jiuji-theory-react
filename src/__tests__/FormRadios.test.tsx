import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import FormRadios from "../components/FormRadios";
import { FormRadiosProps as props } from "../samples/FormRadios.sample";
import { Formik, Form } from "formik";
import "jest-styled-components";

test("should call onChange when clicked", async () => {
  render(
    <Formik initialValues={{ name: "" }} onSubmit={jest.fn()}>
      {({ values }) => (
        <Form>
          <FormRadios {...props} selected={values.name} />
        </Form>
      )}
    </Formik>
  );
  expect(screen.getByText("unchecked").parentElement).toHaveStyleRule(
    "background-color",
    "#fff"
  );

  fireEvent.click(screen.getByText("unchecked"));

  await waitFor(() => {
    expect(screen.getByText("unchecked").parentElement).toHaveStyleRule(
      "background-color",
      "rgba(18,70,246,0.10)"
    );
  });
});

test("should render a checked radio button", () => {
  const { asFragment } = render(
    <Formik initialValues={{ name: "" }} onSubmit={jest.fn()}>
      <Form>
        <FormRadios {...props} selected="checked" />
      </Form>
    </Formik>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("should render an unchecked radio button", () => {
  const { asFragment } = render(
    <Formik initialValues={{ name: "" }} onSubmit={jest.fn()}>
      <Form>
        <FormRadios {...props} selected="unchecked" />
      </Form>
    </Formik>
  );
  expect(asFragment()).toMatchSnapshot();
});
