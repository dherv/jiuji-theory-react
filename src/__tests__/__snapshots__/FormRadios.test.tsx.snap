// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a checked radio button 1`] = `
<DocumentFragment>
  .c0 {
  display: block;
  margin-bottom: 0.5rem;
  font-weight: bold;
  font-size: 0.75rem;
}

.c2 {
  padding: 0.5rem 1rem;
  font-size: 0.75rem;
  font-weight: 600;
  color: #1246F6;
  background-color: rgba(18,70,246,0.10);
  border: 1px solid transparent;
  -webkit-transition: 0.5s;
  transition: 0.5s;
  padding: 0;
  margin-right: 1rem;
}

.c3 {
  padding: 0.5rem 1rem;
  font-size: 0.75rem;
  font-weight: 600;
  color: rgba(18,70,246,0.5);
  background-color: #fff;
  border: 1px solid rgba(18,70,246,0.5);
  -webkit-transition: 0.5s;
  transition: 0.5s;
  padding: 0;
  margin-right: 1rem;
}

.c3:hover {
  background-color: rgba(18,70,246,0.10);
  border: 1px solid transparent;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<form
    action="#"
  >
    <div
      aria-labelledby="name"
      role="group"
    >
      <div
        class="c0"
        id="name"
      >
        name
      </div>
      <div
        class="c1"
      >
        <button
          class="sc-AxjAm c2"
          type="button"
        >
          <label
            for="checked"
            style="display: block; padding: .5rem 1rem;"
          >
            checked
          </label>
        </button>
        <input
          id="checked"
          name="name"
          style="display: none;"
          type="radio"
          value="checked"
        />
        <button
          class="sc-AxjAm c3"
          type="button"
        >
          <label
            for="unchecked"
            style="display: block; padding: .5rem 1rem;"
          >
            unchecked
          </label>
        </button>
        <input
          id="unchecked"
          name="name"
          style="display: none;"
          type="radio"
          value="unchecked"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`should render an unchecked radio button 1`] = `
<DocumentFragment>
  .c0 {
  display: block;
  margin-bottom: 0.5rem;
  font-weight: bold;
  font-size: 0.75rem;
}

.c2 {
  padding: 0.5rem 1rem;
  font-size: 0.75rem;
  font-weight: 600;
  color: rgba(18,70,246,0.5);
  background-color: #fff;
  border: 1px solid rgba(18,70,246,0.5);
  -webkit-transition: 0.5s;
  transition: 0.5s;
  padding: 0;
  margin-right: 1rem;
}

.c2:hover {
  background-color: rgba(18,70,246,0.10);
  border: 1px solid transparent;
}

.c3 {
  padding: 0.5rem 1rem;
  font-size: 0.75rem;
  font-weight: 600;
  color: #1246F6;
  background-color: rgba(18,70,246,0.10);
  border: 1px solid transparent;
  -webkit-transition: 0.5s;
  transition: 0.5s;
  padding: 0;
  margin-right: 1rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<form
    action="#"
  >
    <div
      aria-labelledby="name"
      role="group"
    >
      <div
        class="c0"
        id="name"
      >
        name
      </div>
      <div
        class="c1"
      >
        <button
          class="sc-AxjAm c2"
          type="button"
        >
          <label
            for="checked"
            style="display: block; padding: .5rem 1rem;"
          >
            checked
          </label>
        </button>
        <input
          id="checked"
          name="name"
          style="display: none;"
          type="radio"
          value="checked"
        />
        <button
          class="sc-AxjAm c3"
          type="button"
        >
          <label
            for="unchecked"
            style="display: block; padding: .5rem 1rem;"
          >
            unchecked
          </label>
        </button>
        <input
          id="unchecked"
          name="name"
          style="display: none;"
          type="radio"
          value="unchecked"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;
